File: JPP/lab-1/z1/recu.c
#include "lib.h"

// PRIVATE

void swap(int* a, int* b)
{
    const int temp = *a;
    *a = *b;
    *b = temp;
}

int gcdr(int a, int b)
{
    return b == 0 ? a : gcdr(b, a % b);
}

Pair egcd(int a, int b)
{
    Pair result = {1,0};
    
    if (b != 0)
    {
        result = egcd(b, a % b);
        const int temp = result.x;

        result.x = result.y;
        result.y = temp - (a / b) * result.y;
    }

    return result;
}

// PUBLIC

int factorial(const int n)
{
    return n < 2 ? 1 : factorial(n - 1) * n;
}

int gcd(int a, int b)
{
    if (a < b)
        swap(&a, &b);

    return gcdr(a, b);
}

Pair solve(int a, int b, int c)
{
    // ax + by = c, gcd(a,b) | c
    // egcd(a,b) -> s,t
    // as + bt = gcd(a,b)
    // gcd(a,b) * k = c = ask + btk
    // x = sk, y = tk
    Pair result = {0,0};

    const int d = gcd(a, b);

    if (c % d == 0)
    {
        const Pair st = egcd(a,b);

        const int s = st.x;
        const int t = st.y;
        const int k = c / d;

        result.x = s*k;
        result.y = t*k;
    }

    return result;
}

File: JPP/lab-1/z7/src/recu.rs
pub mod lib {
    pub struct Pair {
        pub x: i32,
        pub y: i32,
    }

    pub fn factorial(n: i32) -> i32 {
        if n < 2 {
            return 1;
        } else {
            return n * factorial(n - 1);
        }
    }

    fn gcdr(a: i32, b: i32) -> i32 {
        if a == 0 {
            return b;
        } else {
            return gcdr(b % a, a);
        }
    }

    pub fn gcd(mut a: i32, mut b: i32) -> i32 {
        if a < b {
            std::mem::swap(&mut a, &mut b);
        }

        return gcdr(a, b);
    }

    pub fn solve(a: i32, b: i32, c: i32) -> Pair {
        let mut result = Pair {x: 0, y: 0};

        let d = gcd(a, b);

        if c % d == 0 {
            result = egcd(a, b);
            result.x *= c / d;
            result.y *= c / d;
        }

        return result;
    }

    fn egcd(a: i32, b: i32) -> Pair {
        let mut result = Pair {x: 1, y: 0};

        if b != 0 {
            result = egcd(b, a % b);
            let temp = result.x;

            result.x = result.y;
            result.y = temp - (a / b) * result.y;
        }

        return result;
    }
}